pkg/apis/options/options.go

```diff
@@ -58,6 +58,8 @@ type Options struct {
 	SSLInsecureSkipVerify bool     `flag:"ssl-insecure-skip-verify" cfg:"ssl_insecure_skip_verify"`
 	SkipAuthPreflight     bool     `flag:"skip-auth-preflight" cfg:"skip_auth_preflight"`
 	ForceJSONErrors       bool     `flag:"force-json-errors" cfg:"force_json_errors"`
+	// Force PKCE Code Challenges if method is not detected via RFC-8414 metadata document
+	ForceCodeChallengeMethod string `flag:"force-code-challenge-method" cfg:"force_code_challenge_method"`
```

Comment:
Is PKCE supported on all providers? Are we going to need some way to disable this (or rather make it opt-in)?

Reply:
I explained this in the commit message. No PKCE is not supported on all platforms, this is currently opt-in with the exception of if the provider specifically expresses which code challenge methods they support in their discovery document.
If a provider states they support PKCE, we turn it on. Google, Dex, Auth0 advertise this capability in their discovery document. Azure supports it, but does not implement the spec which displays this in their discovery document so oauth2-proxy would default to PKCE off to be safe (even though the spec for OAuth says to ignore any extra URL params, this isn't always the case when implementing). This command line argument allows you to enable it even if the provider doesn't express support.

The only case that is disputed so far is Okta, which according to comments claims there's an error with providing PKCE, even though they express support in their discovery document. I've asked in Slack if there's anyway this can be verified since it's not clear based on the comments if this is due to misconfiguration (lots of discussion about not providing a client secret, and error seems to indicate misconfiguration)

Reply:
I added my similar apprehensions here: https://github.com/oauth2-proxy/oauth2-proxy/pull/1541#discussion_r810549174

I think until we are in a major release, this should explicitly be an "OPT-IN" feature (no auto-detection via discovery documents by default). I think this should bake for at least a minor version or two until we consider switching it to on by default.